### DISCUSSION TITLE
VERCEL_URL not accurate on every domains

### DISCUSSION THREAD

**Discussion by Brandfetch** [vercel/vercel#5230] [Help] [answered]
I'm using `VERCEL_URL` to get the current hostname. The issue is that it only works with some domains in the vercel plateform such as production or some preview links. 

These will return the correct URL.
_[project_name]_[build_uid].vercel.app_ (e.g. **google_e31fko4js.vercel.app**)

These won't.
_[domain_name].[suffix]_ (e.g. **google.com**)
_[project_name].[project_name]_.vercel.app_ (**google.google.vercel.app** returns google_e31fko4js.vercel.app)
_[project_name].vercel.app_ (**google.vercel.app** returns google_e31fko4js.vercel.app)
_[project_name]-git-master.[project_name].vercel.app_ (**google-git-master.google.vercel.app** returns google_e31fko4js.vercel.app)

It's confusing, since I can't get the correct URL my app does not work on these builds which make them useless.

What would be the correct NextJS/Vercel method to get the current & correct hostname?

**Comment by styfle** [comment, marked as answer]
Hi @Brandfetch 

`VERCEL_URL` is not the domain, its the unique deployment URL as you have discovered.

If you're using a serverless function, you can read the [`host`](https://vercel.com/docs/edge-network/headers#host) header, such as `req.headers.host` with Node.js during runtime.

If your deployment is static, you could instead add an [environment variable](https://vercel.com/docs/build-step#environment-variables) with your expected domain to read it at build time.

**Comment by DmiPet** [comment]
Hi @styfle,

Do you know of any reliable way to handle this dynamically, specifically for preview deployments that are made on branches, pull requests, etc.?

**Comment by jasonkuhrt** [comment]
Reading the `host` header will probably work for our use-case however its interesting to wonder about if an environment variable matching the domain currently under use for the request should be possible in a serverless environment where each request in theory is a fresh invocation. Of course in practice it is not, multiple invocations do share state on most/all (?) platforms, and so this is probably not doable, and the only reliable way in the end is reading the host header.

**Comment by achen187** [comment]
Is there any update on this? @styfle could you elaborate on your answer, I don't quite understand how that would work.

How would adding an environment variable solve this issue?

**Comment by stevenmunro** [comment]
I realised this today, when a scheduling service I set up attempted to call back the website, it was calling a previous build because VERCEL_URL changes every time.

VERCEL_PRIMARY_DOMAIN would be handy here.

**Reply by leerob** [reply]
These docs might be helpful for you 😄  https://vercel.com/docs/concepts/deployments/automatic-urls
